pam/pam.go

```diff
@@ -69,7 +82,23 @@ func sendReturnMessageToPam(mTx pam.ModuleTransaction, retStatus adapter.PamRetu
 
 // Authenticate is the method that is invoked during pam_authenticate request.
 func (h *pamModule) Authenticate(mTx pam.ModuleTransaction, flags pam.Flags, args []string) error {
-	return h.handleAuthRequest(authd.SessionMode_AUTH, mTx, flags, args)
+	// Do not try to start authentication again if we've been already through this.
+	_, err := mTx.GetData(authenticatedKey)
+	if err == nil && parseArgs(args)["force_reauth"] != "true" {
+		return pam.ErrIgnore
+	}
+	if err != nil && !errors.Is(err, pam.ErrNoModuleData) {
+		return err
+	}
+
+	err = h.handleAuthRequest(authd.SessionMode_AUTH, mTx, flags, args)
+
+	if err == nil || errors.Is(err, pam.ErrIgnore) {
+		if err := mTx.SetData(authenticatedKey, struct{}{}); err != nil {
+			return err
+		}
+	}
+	return err
```

Comment:
Remember about the happy path...

```Go
if err != nil && !errors.Is(err, pam.ErrIgnore) {
    return err        
}

return mTx.SetData(authenticatedKey, struct{}{})
```

Reply:
Yeah, but that would also happen on your previous logic so it's not the happy path's fault 😉

Reply:
Nope, it wasn't (in fact CI wasn't fail as per https://github.com/ubuntu/authd/actions/runs/8069328331/job/22044497208?pr=226)

Since in case of ignore error it was returning the one on outer scope. However, I'm happier with the happy path too :)

Reply:
Oh, interesting... So if we `if err := ...` it already scopes the value and doesn't shadow the previous variable. Good to know.